net/flowtrack/flowtrack.go

```diff
@@ -11,23 +11,69 @@ package flowtrack
 
 import (
 	"container/list"
+	"encoding/json"
 	"fmt"
 	"net/netip"
 
 	"tailscale.com/types/ipproto"
 )
 
+// MakeTuple makes a Tuple out of netip.AddrPort values.
+func MakeTuple(proto ipproto.Proto, src, dst netip.AddrPort) Tuple {
+	return Tuple{
+		Proto:   proto,
+		Src:     src.Addr().As16(),
+		SrcPort: src.Port(),
+		Dst:     dst.Addr().As16(),
+		DstPort: dst.Port(),
+	}
+}
+
 // Tuple is a 5-tuple of proto, source and destination IP and port.
+//
+// This struct originally used netip.AddrPort, but that was about twice as slow
+// when used as a map key due to the alignment and extra space for the IPv6 zone
+// pointers (unneeded for all our current 2024-06-17 flowtrack needs).
+//
+// This struct is packed optimally and doesn't contain gaps or pointers.
 type Tuple struct {
-	Proto ipproto.Proto  `json:"proto"`
-	Src   netip.AddrPort `json:"src"`
-	Dst   netip.AddrPort `json:"dst"`
+	Src     [16]byte
+	Dst     [16]byte
+	SrcPort uint16
+	DstPort uint16
+	Proto   ipproto.Proto
+}
+
+func (t Tuple) SrcAddr() netip.Addr {
+	return netip.AddrFrom16(t.Src).Unmap()
+}
+
+func (t Tuple) DstAddr() netip.Addr {
+	return netip.AddrFrom16(t.Dst).Unmap()
 }
 
 func (t Tuple) String() string {
 	return fmt.Sprintf("(%v %v => %v)", t.Proto, t.Src, t.Dst)
 }
 
+func (t Tuple) MarshalJSON() ([]byte, error) {
```

Comment:
So we need an Unmarshal equivalent?

Reply:
Dunno! I can't actually find anybody using this as JSON. The JSON struct tags made me think it was JSON-ified somewhere, though.

Reply:
Added Unmarshal+test